File: DataStructures/Java/DoublyCircularLinkedList.java
class DoublyCircularLinkedList {
    static class Node {
        // CSOFF: VisibilityModifier
        int data;
        Node next;
        Node prev;
        // CSON: VisibilityModifier

        Node(final int newData) {
            this.data = newData;
            this.next = null;
            this.prev = null;
        }
    }

    private Node head = null;
    private Node tail = null;

    public final void insertAtLast(final int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
            this.head.next = this.head;
            this.head.prev = this.head;
            return;
        }
        this.tail.next = newNode;
        newNode.prev = this.tail;
        newNode.next = this.head;
        this.head.prev = newNode;
        this.tail = newNode;
    }

    public final void insertAtFirst(final int data) {
        Node newNode = new Node(data);
        newNode.next = this.head;
        newNode.prev = this.tail;
        this.head.prev = newNode;
        this.tail.next = newNode;
        this.head = newNode;
    }

    public final void insertAfter(final int data, final int after) {
        Node newNode = new Node(data);
        Node temp = this.head;
        while (temp != null) {
            if (temp.data == after) {
                newNode.next = temp.next;
                newNode.prev = temp;
                temp.next.prev = newNode;
                temp.next = newNode;
                return;
            }
            temp = temp.next;
        }
    }

    public final void delete(final int data) {
        if (this.head == null) {
            return;
        }
        if (this.head.data == data) {
            this.head = this.head.next;
            this.head.prev = this.tail;
            this.tail.next = this.head;
            return;
        }
        Node temp = this.head;
        while (temp != null) {
            if (temp.data == data) {
                temp.prev.next = temp.next;
                temp.next.prev = temp.prev;
                return;
            }
            temp = temp.next;
        }
    }

    public final void delete(final Node node) {
        if (this.head == null) {
            return;
        }
        if (this.head == node) {
            this.head = this.head.next;
            this.head.prev = this.tail;
            this.tail.next = this.head;
            return;
        }
        Node temp = this.head;
        while (temp != null) {
            if (temp == node) {
                temp.prev.next = temp.next;
                temp.next.prev = temp.prev;
                return;
            }
            temp = temp.next;
        }
    }

    public final void print() {
        Node temp = this.head;
        do {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        } while (temp != this.head);
    }

    public static void main(final String[] args) {
        CircularLinkedList list = new CircularLinkedList();
        list.insertAtLast(1);
        list.insertAtLast(2);
        list.insertAtLast(3);
        list.insertAtLast(4);
        list.insertAtLast(5);
        list.insertAtFirst(0);
        list.insertAfter(6, 5);
        list.print();
    }
}

File: DataStructures/C/doubly_circular_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct Node
{
    int data;
    struct Node *next;
    struct Node *prev;
};

struct Node *head = NULL;
struct Node *tail = NULL;

void insert_at_last(int data)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));

    if (head == NULL)
    {
        new_node->data = data;
        new_node->next = new_node;
        new_node->prev = new_node;
        head = new_node;
        tail = new_node;
    }
    else
    {
        new_node->data = data;
        new_node->next = head;
        new_node->prev = tail;
        tail->next = new_node;
        head->prev = new_node;
        tail = new_node;
    }
}

void insert_at_first(int data)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));

    new_node->data = data;

    new_node->next = head;
    new_node->prev = tail;

    head->prev = new_node;
    tail->next = new_node;

    head = new_node;
}

void delete_node(int data)
{
    if (head == NULL)
    {
        printf("List is empty\n");
        return;
    }

    struct Node *temp = head;
    while (temp->data != data)
    {
        temp = temp->next;
        if (temp == head)
        {
            printf("Data not found\n");
            return;
        }
    }
    if (temp == head)
    {
        head = head->next;
    }
    if (temp == tail)
    {
        tail = tail->prev;
    }
    temp->prev->next = temp->next;
    temp->next->prev = temp->prev;
    free(temp);
}

void display()
{
    struct Node *temp = head;
    do
    {
        printf("%d <-> ", temp->data);
        temp = temp->next;
    } while (temp != head);

    printf("\n");
}

int main()
{
    for (int i = 0; i < 10; i++)
    {
        insert_at_last(i);
    }
    display();

    for (int i = 10; i < 20; i++)
    {
        insert_at_first(i);
    }
    display();

    delete_node(19);

    display();

    delete_node(-5);

    display();

    return 0;
}